File: Cpp-C/Ema/Src/Rdm/Impl/EnumTypeTableImpl.cpp
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2017,2019-2020,2025 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

#include "EnumTypeTableImpl.h"
#include "EnumTypeImpl.h"

using namespace refinitiv::ema::access;
using namespace refinitiv::ema::rdm;

EnumTypeTableImpl::EnumTypeTableImpl() :
_pEnumTypeList(0),
_pFidsList(0)
{
}

EnumTypeTableImpl::EnumTypeTableImpl(RsslEnumTypeTable* rsslEnumTypeTable) :
_pEnumTypeTable(rsslEnumTypeTable),
_pEnumTypeList(0),
_pFidsList(0),
_refreshEnumTypeList(false)
{
}

EnumTypeTableImpl::~EnumTypeTableImpl()
{
	if ( _pEnumTypeList )
	{
		delete _pEnumTypeList;	
		_pEnumTypeList = 0;
	}

	if ( _pFidsList )
	{
   		delete _pFidsList;
		_pFidsList = 0;
	}
}

RsslEnumTypeTable* EnumTypeTableImpl::getRsslEnumTypeTable() const
{
	return _pEnumTypeTable;
}

void EnumTypeTableImpl::rsslEnumTypeTable(RsslEnumTypeTable* enumTypeTable)
{
	_refreshEnumTypeList = true;

	_pEnumTypeTable = enumTypeTable;
}

const refinitiv::ema::access::EmaVector<EnumType>& EnumTypeTableImpl::getEnumTypes() const
{
	if (_pEnumTypeList == 0)
	{
		_pEnumTypeList = new EmaVector<EnumType>(_pEnumTypeTable->maxValue);
	}
	else if (_refreshEnumTypeList )
	{
		_pEnumTypeList->clear();
	}

	RsslEnumType*	rsslEnumType = 0;
	EnumType	enumType;

	for (UInt32 index = 0; index <= _pEnumTypeTable->maxValue; index++)
	{
		rsslEnumType = *(_pEnumTypeTable->enumTypes + index);

		if (rsslEnumType)
		{
			enumType._pImpl->rsslEnumType(rsslEnumType);
			_pEnumTypeList->push_back(enumType);
		}
	}

	_refreshEnumTypeList = false;

	return *_pEnumTypeList;
}

const refinitiv::ema::access::EmaVector<refinitiv::ema::access::Int16>& EnumTypeTableImpl::getFidReferences() const
{
	if (_pFidsList == 0)
	{
		_pFidsList = new EmaVector<Int16>(_pEnumTypeTable->fidReferenceCount);
	}
	else
	{
		_pFidsList->clear();
	}

	for (UInt32 index = 0; index < _pEnumTypeTable->fidReferenceCount; index++)
	{
		_pFidsList->push_back(_pEnumTypeTable->fidReferences[index]);
	}

	return *_pFidsList;
}

const refinitiv::ema::access::EmaString& EnumTypeTableImpl::toString() const
{
	_stringToString.set(0, 256);

	RsslEnumType* rsslEnumType = 0;
	EnumType enumType;

	for (UInt32 index = 0; index < _pEnumTypeTable->fidReferenceCount; index++)
	{
		_stringToString.append("(Referenced by Fid ").append(_pEnumTypeTable->fidReferences[index]).append(")\n");
	}

	for (UInt32 index = 0; index <= _pEnumTypeTable->maxValue; index++)
	{
		rsslEnumType = *(_pEnumTypeTable->enumTypes + index);

		if (rsslEnumType)
		{
			enumType._pImpl->rsslEnumType(rsslEnumType);

			_stringToString.append(enumType.toString()).append("\n");
		}
	}
	
	return _stringToString;
}

File: Java/Ema/Core/src/main/java/com/refinitiv/ema/access/EnumTypeTableImpl.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2020,2024 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.ema.access;

import java.util.ArrayList;
import java.util.List;

import com.refinitiv.ema.rdm.EnumType;
import com.refinitiv.ema.rdm.EnumTypeTable;
import com.refinitiv.eta.valueadd.common.VaNode;
import com.refinitiv.eta.valueadd.common.VaPool;

class EnumTypeTableImpl extends VaNode implements EnumTypeTable {
	
	private com.refinitiv.eta.codec.EnumTypeTable 		rsslEnumTypeTable;
	private List<EnumType> 									enumTypeList;
	private List<Integer> 									fidsList;
	private StringBuilder									toStringValue;
	private boolean 										refreshEnumTypeList;
	
	private  VaPool											enumTypePool = new VaPool(false);
	
	EnumTypeTableImpl() {
		
	}

	EnumTypeTableImpl enumTypeTable(com.refinitiv.eta.codec.EnumTypeTable enumTypeTable) {
		
		refreshEnumTypeList = true;
		
		rsslEnumTypeTable = enumTypeTable;
		
		return this;
	}
	
	EnumTypeTableImpl clear()
	{
		refreshEnumTypeList = true;
		
		clearEnumTypeList();
		
		if ( fidsList != null )
		{
			fidsList.clear();
		}
		
		return this;
	}
	
	private void clearEnumTypeList()
	{
		if ( enumTypeList != null && enumTypeList.size() != 0 )
		{
			for(int index = 0; index < enumTypeList.size(); index++ )
			{
				((EnumTypeImpl)enumTypeList.get(index)).returnToPool();
			}
			
			enumTypeList.clear();
		}
	}

	@Override
	public List<EnumType> enumTypes() {
		
		if ( enumTypeList == null )
		{
			enumTypeList = new ArrayList<>(rsslEnumTypeTable.maxValue());
		}
		else if ( refreshEnumTypeList )
		{
			clearEnumTypeList();
		}
		else
		{
			return enumTypeList;
		}
		
		com.refinitiv.eta.codec.EnumType rsslEnumType;
		
		for(int index = 0; index <= rsslEnumTypeTable.maxValue(); index++ )
		{
			rsslEnumType = rsslEnumTypeTable.enumTypes()[index];
			
			if ( rsslEnumType != null )
			{
				enumTypeList.add(getEnumType(rsslEnumType));
			}
		}
		
		refreshEnumTypeList = false;
		
		return enumTypeList;
	}

	@Override
	public List<Integer> FidReferences() {
		
		if ( fidsList == null )
		{
			fidsList = new ArrayList<>(rsslEnumTypeTable.fidReferenceCount());
		}
		else
		{
			fidsList.clear();
		}
		
		for(int index = 0; index < rsslEnumTypeTable.fidReferenceCount(); index++)
		{
			fidsList.add(Integer.valueOf(rsslEnumTypeTable.fidReferences()[index]));
		}
		
		return fidsList;
	}
	
	@Override
	public String toString() {
	
		if ( toStringValue == null )
		{
			toStringValue = new StringBuilder(256);
		}
		else
		{
			toStringValue.setLength(0);
		}
		
		for(int index = 0; index < rsslEnumTypeTable.fidReferenceCount(); index++)
		{
			toStringValue.append("(Referenced by Fid ").append(rsslEnumTypeTable.fidReferences()[index]).append(")\n");
		}
		
		com.refinitiv.eta.codec.EnumType rsslEnumType;
		EnumTypeImpl enumTypImpl = new EnumTypeImpl();
		
		for(int index = 0; index <= rsslEnumTypeTable.maxValue(); index++ )
		{
			rsslEnumType = rsslEnumTypeTable.enumTypes()[index];
			
			if ( rsslEnumType != null )
			{
				toStringValue.append(enumTypImpl.enumType(rsslEnumType)).append("\n");
			}
		}
		
		return toStringValue.toString();
	}
	
	private EnumTypeImpl getEnumType(com.refinitiv.eta.codec.EnumType enumType)
	{
		EnumTypeImpl enumTypeImpl = (EnumTypeImpl)enumTypePool.poll();
		
		if ( enumTypeImpl == null )
		{
			enumTypeImpl = new EnumTypeImpl();
			enumTypePool.updatePool(enumTypeImpl);
		}
	
		enumTypeImpl.enumType(enumType);
	
		return enumTypeImpl;
	}
}
